qiskit_experiments/framework/backend_timing_mixin.py

```diff
@@ -0,0 +1,169 @@
+        are needed to get to the next valid sample for the start of a pulse in
+        a subsequent instruction.
+
+        This method is useful when it is desired that a delay instruction
+        accurately reflect how long the hardware will pause between two other
+        instructions.
+
+        Args:
+            time: The nominal delay time to convert
+            next_instruction: Either "pulse", "acquire", or "both" to indicate
+                whether the next instruction will be a pulse, an acquire
+                instruction, or both.
+
+        Returns:
+            The delay duration in samples or seconds, depending on the value of
+            :meth:`.BackendTimingMixin._delay_unit`.
+
+        Raises:
+            QiskitError: Bad value for ``next_instruction``.
+        """
+        if self._delay_unit == "s":
+            return time
+
+        timing_constraints = getattr(self.backend.configuration(), "timing_constraints", {})
+        pulse_alignment = timing_constraints.get("pulse_alignment", 1)
+        acquire_alignment = timing_constraints.get("acquire_alignment", 1)
+
+        if next_instruction == "both":
+            # We round the delay values to the least common multiple of the
+            # alignment constraints so that the delay can fill all the time
+            # until a subsequent pulse satisfying the alignment constraints.
+            #
+            # Replace with math.lcm(pulse_alignment, acquire_alignment) when
+            # dropping support for Python 3.8
+            granularity = (
+                pulse_alignment * acquire_alignment // math.gcd(pulse_alignment, acquire_alignment)
+            )
+        elif next_instruction == "pulse":
+            granularity = pulse_alignment
+        elif next_instruction == "acquire":
+            granularity = acquire_alignment
+        else:
+            raise QiskitError(f"Bad value for next_instruction: {next_instruction}")
+
+        samples = int(round(time / self._dt / granularity) * granularity)
+
+        return samples
+
+    def _pulse_duration(self, time: float) -> int:
+        """The number of samples giving a valid pulse duration closest to ``time``
+
+        Args:
+            time: Pulse duration in seconds
+
+        Returns:
+            The number of samples corresponding to ``time``
+
+        Raises:
+            QiskitError: The backend timing constraints' min_length is not a
+                multiple of granularity
+        """
+        timing_constraints = getattr(self.backend.configuration(), "timing_constraints", {})
+        granularity = timing_constraints.get("granularity", 1)
+        min_length = timing_constraints.get("min_length", 1)
+
+        samples = int(round(time / self._dt / granularity)) * granularity
+        samples = max(samples, min_length)
+
+        if min_length % granularity != 0:
+            raise QiskitError("Backend timing does not match assumptions!")
```

Comment:
I'm not sure if this is really tested on front end. Seems like this is backend issue because a user cannot really set these constraints (technically it's possible though).

Reply:
I put this here because I was not sure if it was safe to assume that this is true. Do you think it is? It would make sense if it were true. I was thinking of a weird case where the minimum length were 70 samples and then longer pulses had to be 70 + N*16 samples long for example. If we can assume that min_length == N*granularity so every pulse has to be N*granularity, then I can take this out.

Reply:
I guess this condition is generally satisfied. For example [Keysight AWG spec](https://www.keysight.com/us/en/assets/7018-06341/data-sheets/5992-3361.pdf). I remember we just ported these words from convention in this field. I'm not sure `min_length` indicates whether the minimum "valid" length or just some minimum random number, but I think it is former.

Reply:
Okay, I removed this check.